phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/PhantomJSEnv.scala

```diff
@@ -215,50 +216,33 @@ final class PhantomJSEnv(config: PhantomJSEnv.Config) extends JSEnv {
         <title>Phantom.js Launcher</title>
         <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />""")
 
-    /* The most horrible way to identify the test script of
-     * `RunTests.syntaxErrorTest` from the conformance test suite. The script
-     * we are looking for contains whitespace and exactly one '{'.
-     * Part of a workaround for #10.
-     */
-    def isSyntaxErrorTestScript(script: MemVirtualBinaryFile): Boolean = {
-      script.path == "testScript.js" && {
-        val is = script.inputStream
-        try {
-          // Because it is a MemVirtualBinaryFile
-          assert(is.isInstanceOf[ByteArrayInputStream])
-
-          val buf = new Array[Byte](128)
-          val n = is.read(buf)
-          n > 0 && n < buf.length && {
-            assert(is.read() < 0) // because it is a ByteArrayInputStream
-            (0 until n).forall { i =>
-              val b = buf(i)
-              b == ' ' || b == '\n' || b == '{'
-            } && {
-              (0 until n).count(i => buf(i) == '{') == 1
-            }
-          }
-        } finally {
-          is.close()
-        }
-      }
+    // Enhance scripts with an awesome hack to detect syntax errors
+    val scriptsAndSyntaxErrorHacks = scripts.zipWithIndex.flatMap {
+      case (script, index) =>
+        val hackedScriptOutputStream = new ByteArrayOutputStream()
+        Utils.pipeInputStreamToOutputStream(script.inputStream,
+            hackedScriptOutputStream)
+        val hackedScriptCode =
+          s"\n;\nvar SCALAJS_PHANTOMJS_SYNTAXERROR_HACK_$index = true;\n"
+        hackedScriptOutputStream.write(hackedScriptCode.getBytes(UTF_8))
+        val hackedScript = MemVirtualBinaryFile(script.path,
+            hackedScriptOutputStream.toByteArray())
+
+        val checkScript = MemVirtualBinaryFile.fromStringUTF8(
+            s"checkSyntaxError$index.js",
+            s"""
+              |if (typeof SCALAJS_PHANTOMJS_SYNTAXERROR_HACK_$index === 'undefined')
+              |  throw new SyntaxError("Syntax error in ${escapeJS(script.path)}");
+            """.stripMargin)
+
+        List(hackedScript, checkScript)
     }
 
-    for (script <- scripts) {
-      script match {
-        /* The most horrible way to allow RunTests.syntaxErrorTest in the
-         * conformance test suite to succeed, despite #10.
-         */
-        case script: MemVirtualBinaryFile if isSyntaxErrorTestScript(script) =>
-          out.write(
-              """<script type="text/javascript">throw new SyntaxError("syntax error");</script>""" + "\n")
-
-        case _ =>
-          val scriptURI = materialize(script)
-          val fname = htmlEscape(fixFileURI(scriptURI).toASCIIString)
-          out.write(
-              s"""<script type="text/javascript" src="$fname"></script>""" + "\n")
-      }
+    for (script <- scriptsAndSyntaxErrorHacks) {
+      val scriptURI = materialize(script)
```

Comment:
I think with these changes, `materialize` always takes `MemVirtualBinaryFile`. Maybe simplify it in light of that?